src/ext/Util/ca/scaexec.cpp

```diff
@@ -747,95 +820,136 @@ extern "C" UINT __stdcall CreateUser(
+
+                if (*pwzScriptKey)
                 {
-                    // Change the existing user's password and attributes again then try
-                    // to update user with this new data
-                    SetUserPasswordAndAttributes(puserInfo, pwzPassword, iAttributes);
+                    // Try to open the rollback script
+                    hr = WcaCaScriptOpen(WCA_ACTION_INSTALL, WCA_CASCRIPT_ROLLBACK, FALSE, pwzScriptKey, &hRollbackScript);
+
+                    if (INVALID_HANDLE_VALUE != hRollbackScript)
+                    {
+                        WcaCaScriptClose(hRollbackScript, WCA_CASCRIPT_CLOSE_PRESERVE);
+                    }
+                    else
+                    {
+                        hRollbackScript = NULL;
+                        hr = WcaCaScriptCreate(WCA_ACTION_INSTALL, WCA_CASCRIPT_ROLLBACK, FALSE, pwzScriptKey, FALSE, &hRollbackScript);
+                        ExitOnFailure(hr, "Failed to open rollback CustomAction script.");
+
+                        iRollbackAttributes = 0;
+                        hr = GetExistingUserRightsAssignments(pwzDomain, pwzName, &iOriginalAttributes);
+                        if (FAILED(hr))
+                        {
+                            WcaLogError(hr, "failed to get existing user rights: %ls, continuing anyway.", pwzName);
+                        }
+                        else
+                        {
+                            if (!(SCAU_ALLOW_LOGON_AS_SERVICE & iOriginalAttributes) && (SCAU_ALLOW_LOGON_AS_SERVICE & iAttributes))
+                            {
+                                iRollbackAttributes |= SCAU_ALLOW_LOGON_AS_SERVICE;
+                            }
+
+                            if (!(SCAU_ALLOW_LOGON_AS_BATCH & iOriginalAttributes) && (SCAU_ALLOW_LOGON_AS_BATCH & iAttributes))
+                            {
+                                iRollbackAttributes |= SCAU_ALLOW_LOGON_AS_BATCH;
+                            }
+                        }
+
+                        hr = WcaCaScriptWriteString(hRollbackScript, pUserInfo1->usri1_comment);
+                        ExitOnFailure(hr, "Failed to add rollback comment to rollback script.");
+
+                        if (!pUserInfo1->usri1_comment || !*pUserInfo1->usri1_comment)
+                        {
+                            iRollbackAttributes |= SCAU_REMOVE_COMMENT;
+                        }
+
+                        hr = WcaCaScriptWriteNumber(hRollbackScript, iRollbackAttributes);
+                        ExitOnFailure(hr, "Failed to add rollback attributes to rollback script.");
+
+                        // Nudge the system to get all our rollback data written to disk.
+                        WcaCaScriptFlush(hRollbackScript);
+                    }
+                }
+
+                ExitOnNetworkError(::SetUserPassword(pwzServerName, pwzName, pwzPassword));
```

Comment:
If these are best effort, is it better to at least try the other `SetXxx` functions even if one fails? Also, it would probably be a good idea to log when the user info is not updated, especially if it also says", continuing anyway.".

Reply:
I've taken a shot at this. I may have erred on the side of leniency. Make sure that I did not allow the process to continue
when it should not.